File: Server/ChatApp/ChatApp.Backend/Core/Common/Result.cs
namespace ChatApp.Backend.Core.Common;

public record Result<T>(
    bool IsSuccess,
    T? Data = default,
    string? ErrorMessage = null,
    int StatusCode = StatusCodes.Status200OK
)
{
    public static Result<T> Success(T data, int statusCode = StatusCodes.Status200OK) =>
        new(true, data, null, statusCode);

    public static Result<T> Failure(
        string errorMessage,
        int statusCode = StatusCodes.Status400BadRequest
    ) => new(false, default, errorMessage, statusCode);
}

File: Front/lib/core/result_type/result.dart
class Result<T> {
  Result._({required this.isSuccess, this.errorMessage, this.data});

  factory Result.success(T data) => Result._(isSuccess: true, data: data);

  factory Result.failure(String errorMessage) =>
      Result._(isSuccess: false, errorMessage: errorMessage);

  final bool isSuccess;
  final String? errorMessage;
  final T? data;
}
